examples/ode-integrator.dx

```diff
@@ -1,60 +1,179 @@
+  safety = 0.9
+  ifactor = 10.
+  dfactor = 0.2
+  order = 5.
+  dfactor = select (mean_error_ratio < 1.) 1. dfactor
+  err_ratio = sqrt mean_error_ratio
+  minfac = min ( (pow err_ratio (1. / order)) / safety) (1. / dfactor)
+  factor = max (1. / ifactor) minfac
+  select (mean_error_ratio == 0.) (last_step * ifactor) (last_step / factor)
+
+
+def odeint (func: d=>Float -> Time -> d=>Float)
+           (z0: d=>Float) (t0: Time) (times: n=>Time) : n=>d=>Float =
+  -- Adaptive stepsize (Dormand-Prince) Runge-Kutta odeint implementation.
+  --  Args:
+  --    func: time derivative of the solution z at time t.
+  --    z0: the initial value for the state.
+  --    t: times for evaluation. values must be strictly increasing.
+  --  Returns:
+  --    Values of the solution at each time point in times.
+  rtol = 1.4e-8 -- relative local error tolerance for solver.
+  atol = 1.4e-8 -- absolute local error tolerance for solver.
+  max_iters = 10000
+
+  integrate_to_next_time = \iter init_carry.
+    target_t = times.iter
+
+    stopping_condition = \(_, _, t, dt, _, _).
+      -- State of solver: (next state, next f, next time, dt, t, interp coeffs)
+      -- def State (v:Type) : Type = (v & v & Time & Time & Time & (Fin 5)=>v)
+      -- This ended up being unnecessary to spell anywhere, but was
+      -- useful for debugging.
+      (t < target_t) && (dt > 0.0) && (ordinal iter < max_iters)
+
+    possible_step = \(z, f, t, dt, last_t, interp_coeff).
+      (next_z, next_f, next_z_error, k) = runge_kutta_step func z f t dt
+      next_t = t + dt
+      ratio = error_ratio next_z_error rtol atol z next_z
+      new_interp_coeff = interp_fit_dopri z next_z k dt
+      new_dt = optimal_step_size dt ratio
+
+      move_state = (next_z, next_f, next_t, new_dt,      t, new_interp_coeff)
+      stay_state = (     z,      f,      t, new_dt, last_t,     interp_coeff)
+      select (ratio <= 1.0) move_state stay_state
+
+    -- Take steps until we pass target_t
+    new_state = snd $ withState init_carry \state.
+      while (\(). stopping_condition (get state)) \().
+        state := possible_step (get state)
+    (_, _, t, _, last_t, interp_coeff) = new_state
+
+    -- Interpolate to the target time.
+    relative_output_time = (target_t - last_t) / (t - last_t)
+    z_target = evalpoly interp_coeff relative_output_time
+    (new_state, z_target)
+
+  f0 = func z0 t0
+  init_step = initial_step_size func t0 z0 4 rtol atol f0
+  init_interp_coeff = zero  -- dummy vals
+  init_carry = (z0, f0, t0, init_step, t0, init_interp_coeff)
+  snd $ scan init_carry integrate_to_next_time
+
 
 '#### Example: Linear dynamics
 
-def myDyn (dict: VSpace a) ?=> (z : a) -> (t:Time) : a = z
+def myDyn (z : a) -> (t:Time) : a = z
 
 z0 = [1.0]
 t0 = 0.0
 t1 = [1.0]
-dt = 0.001
 
-:p odeint myDyn z0 t0 t1 dt
-> [[2.7169456]]
+approx_e = odeint myDyn z0 t0 t1
+:p approx_e
+> [[2.7201762]]
+
+exact_e = [[exp 1.0]]
+
+:p (approx_e - exact_e) -- amount of numerical error
```

Comment:
nit: we could use `~~` now!

Reply:
I would, except I wanted to show the amount of error.  Actually, what I should really show is a graph of error vs tolerance, to show the convergence rate.

Reply:
Ohh that would cool!